### PR TITLE
ENH: numpy histogram bin edges in cut (GH 14627)

### LINKED ISSUE #14627: numpy 1.11.0 nonuniform binning
`numpy 1.11.0` allows for several strings as inputs to create non-uniform binning. It would be useful if this could be implemented in pandas.

[See the `numpy.histogram` docs.](https://docs.scipy.org/doc/numpy/reference/generated/numpy.histogram.html)

### PR DESCRIPTION
**Pull request by blalterman**
Passig a string to `pd.cut` bins kwarg dispatches bin calculation
    to `np.histogram_bin_edges`.

- [X] closes #14627
- [X] tests added / passed
- [X] added whatsnew entry

### CODE REVIEW

File: pandas/core/reshape/tile.py [hunk 1]
```diff
@@ -181,14 +183,37 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
...
+                                        (0.55, 0.775] < (0.775, 1.0]]
     """
     # NOTE: this binning code is changed a bit from histogram for var(x) == 0
 
     # for handling the cut for datetime and timedelta objects
     x_is_series, series_index, name, x = _preprocess_for_cut(x)
     x, dtype = _coerce_to_type(x)
 
-    if not np.iterable(bins):
+    if isinstance(bins, str):
```
Comment (gfyoung) [inline pandas/core/reshape/tile.py]: Use `compat.string_types` instead of `str`.

File: pandas/tests/reshape/test_tile.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 from pandas.core.algorithms import quantile
 import pandas.core.reshape.tile as tmod
 
+import pdb
```
Comment (gfyoung) [inline pandas/tests/reshape/test_tile.py]: You don't need this, I presume?

File: pandas/tests/reshape/test_tile.py [hunk 3]
```diff
@@ -37,6 +38,24 @@ def test_bins(self):
         tm.assert_almost_equal(bins, np.array([0.1905, 3.36666667,
                                                6.53333333, 9.7]))
 
+    def test_str_bins(self):
+        data = np.array([0.1, 0.1, 0.2, 0.5, 0.5, 0.9, 1.0])
```
Comment (gfyoung) [inline pandas/tests/reshape/test_tile.py]: Add comment referencing the issue number above this.

File: pandas/core/reshape/tile.py [hunk 4]
```diff
@@ -35,12 +35,14 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
...
     x : array-like
         The input array to be binned. Must be 1-dimensional.
-    bins : int, sequence of scalars, or pandas.IntervalIndex
+    bins : int, str, sequence of scalars, or pandas.IntervalIndex
         The criteria to bin by.
 
         * int : Defines the number of equal-width bins in the range of `x`. The
           range of `x` is extended by .1% on each side to include the minimum
           and maximum values of `x`.
+        * str : Bin calculaton dispatched to `np.histogram_bin_edges`. See that
```
Comment (jreback) [inline pandas/core/reshape/tile.py]: versionadded tag here

File: pandas/core/reshape/tile.py [hunk 5]
```diff
@@ -181,14 +183,37 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
     >>> pd.cut([0, 0.5, 1.5, 2.5, 4.5], bins)
     [NaN, (0, 1], NaN, (2, 3], (4, 5]]
     Categories (3, interval[int64]): [(0, 1] < (2, 3] < (4, 5]]
+
+    Passng a string for `bins` dispatches the bin calculation to numpy's
```
Comment (jreback) [inline pandas/core/reshape/tile.py]: can you say, starting in 0.24

File: pandas/tests/reshape/test_tile.py [hunk 6]
```diff
@@ -37,6 +38,24 @@ def test_bins(self):
...
+        tm.assert_almost_equal(bins_cut, bins_np)
+        tm.assert_almost_equal(np.round(bins_cut, 4),
+                               np.array([0.0991, 0.325, 0.55, 0.775, 1.0]))
+
+        intervals = IntervalIndex.from_breaks(np.round(bins_np, 4),
+                                              closed="right")
+        expected = Categorical(intervals, ordered=True)
+        tm.assert_index_equal(result.categories,
+                              expected.categories)
+
```
Comment (jreback) [inline pandas/tests/reshape/test_tile.py]: can you add a test for passing a str not-accepted by histogram_bin_edges.

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew entry in other enhancements

File: pandas/core/reshape/tile.py (hunk 1 above)
Comment (bla1089) [reply pandas/core/reshape/tile.py]: Addressed in commit f3323f4

File: pandas/tests/reshape/test_tile.py (hunk 2 above)
Comment (bla1089) [reply pandas/tests/reshape/test_tile.py]: Addressed in f3323f4

File: pandas/tests/reshape/test_tile.py (hunk 3 above)
Comment (bla1089) [reply pandas/tests/reshape/test_tile.py]: Addressed in f3323f4

File: pandas/core/reshape/tile.py (hunk 4 above)
Comment (bla1089) [reply pandas/core/reshape/tile.py]: Added in commit f3323f4

File: pandas/core/reshape/tile.py (hunk 5 above)
Comment (bla1089) [reply pandas/core/reshape/tile.py]: Added in commit f3323f4

File: pandas/tests/reshape/test_tile.py (hunk 6 above)
Comment (bla1089) [reply pandas/tests/reshape/test_tile.py]: Added in commit f3323f4

File: pandas/tests/reshape/test_tile.py [hunk 7]
```diff
@@ -37,6 +36,32 @@ def test_bins(self):
...
+        expected = Categorical(intervals, ordered=True)
+        tm.assert_index_equal(result.categories,
+                              expected.categories)
+        
+        
+        # Test that a `bin` string not present in `np.histogram_bin_edges`
+        # throws a ValueError.
+        tm.assert_raises_regex(ValueError, 
+            "'*' is not a valid estimator for `bins`", 
+            cut, data, "bad bins")
```
Comment (gfyoung) [inline pandas/tests/reshape/test_tile.py]: `tm.assert_raises_regex` has been deprecated.  Use `pytest.raises` instead (you might need to adjust the syntax a little).

Comment (bla1089) [reply pandas/tests/reshape/test_tile.py]: Done

**Comment by bla1089** [comment]
Is there anything left to resolve?

**Comment by gfyoung** [comment]
> Is there anything left to resolve?

@bla1089 : We recently had a huge doc-formatting change.  If you could resolve the merge conflict, that would be fantastic.

**Comment by bla1089** [comment]
Not even sure how to do that. The site says they're too complex to solve
online.

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by bla1089** [comment]
Thank you. I won't pretend to know how to do all that.

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by bla1089** [comment]
Can someone help me interpret the Travis CI and Azure failures?

**Comment by gfyoung** [comment]
> Can someone help me interpret the Travis CI and Azure failures?

@bla1089 : When was `histogram_bin_edges` introduced to `numpy` ?  It seems that it's breaking on some of the `numpy` versions.

**Comment by bla1089** [comment]
I believe "auto" was added as an option to `np.histogram` in 1.11
and `histogram_bin_edges` was exposed independent of `histogram` in 1.15.0
or 1.15.1.

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by gfyoung** [comment]
@bla1089 : Hmm...if you could double check when this was introduced, that would be great.  It looks like this feature only applies to certain `numpy` versions.

**Comment by bla1089** [comment]
The documentation is unclear. `np.histogram` docstring indicates
`bins="string"` option was added in 1.11, as do the release notes. (
https://docs.scipy.org/doc/numpy-1.11.0/release.html) The 1.15.0 release
notes indicate that `histogram_bin_edges` was refactors into an independent
function, but there is no similar notice in the `histogram_bin_edges`
docstring.

For numpy version compatibility, I can add a try-except statement that will
call `np.histogram` when `np.histogram_bin_edges` raises an
`AttributeError`. Would that be sufficient?

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by gfyoung** [comment]
@bla1089 : That seems reasonable for now.  Give that a shot!

**Comment by bla1089** [comment]
@gfyoung can you help me fugue out the status here?

**Comment by gfyoung** [comment]
> can you help me fugue out the status here?

@bla1089 : Sorry, I'm not sure I fully understand you here.  Are you referring to debugging the tests?

**Comment by bla1089** [comment]
yes

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by gfyoung** [comment]
Based on the Travis failures, I see a couple of things:

It looks like your version checking of `numpy` isn't quite working, based on this [build](https://travis-ci.org/pandas-dev/pandas/jobs/457042305).  In addition, something doesn't look quite right with your docstring formatting (see this [build](https://travis-ci.org/pandas-dev/pandas/jobs/457042306)).

Not 100% sure off the top of my head how to fix the doc issues though...

cc @datapythonista

File: pandas/core/reshape/tile.py [hunk 8]
```diff
@@ -181,14 +187,45 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
...
+    `histogram_bin_edges`. (Starting in version 0.24.)
+    >>> pd.cut(array([0.1, 0.1, 0.2, 0.5, 0.5, 0.9, 1.0]),
```
Comment (datapythonista) [inline pandas/core/reshape/tile.py]: leave a blank line before this line

File: pandas/core/reshape/tile.py [hunk 9]
```diff
@@ -181,14 +187,45 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
...
+    ...        bins="auto")
+    ... # doctest: +ELLIPSIS`
```
Comment (datapythonista) [inline pandas/core/reshape/tile.py]: this should go in the previous line, after the code

**Review by datapythonista — COMMENTED** [review]
couple of comments about the docstring, it'd be good to render it to see how it's rendered

File: doc/source/whatsnew/v0.24.0.rst [hunk 10]
```diff
@@ -233,6 +233,7 @@ Other Enhancements
 - :class:`DatetimeIndex` gained :attr:`DatetimeIndex.timetz` attribute. Returns local time with timezone information. (:issue:`21358`)
 - :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support an ``ambiguous`` argument for handling datetimes that are rounded to ambiguous times (:issue:`18946`)
 - :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support a ``nonexistent`` argument for handling datetimes that are rounded to nonexistent times. See :ref:`timeseries.timezone_nonexistent` (:issue:`22647`)
+- :func: `~cut` `bins` kwarg now accepts a string, which is dispatched to `numpy.histogram_bin_edges`. (:issue:`14627`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.24.0.rst]: No space after :func:.

You can link to the numpy method with :func:\`numpy.historgram_bin_edges\`.

File: pandas/core/reshape/tile.py [hunk 11]
```diff
@@ -44,6 +46,10 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
         * sequence of scalars : Defines the bin edges allowing for non-uniform
           width. No extension of the range of `x` is done.
         * IntervalIndex : Defines the exact bins to be used.
+        * str : Bin calculaton dispatched to `np.histogram_bin_edges`. See that
```
Comment (TomAugspurger) [inline pandas/core/reshape/tile.py]: Again, link to the numpy function here with :func:`numpy.histogram_bin_edges`.

File: pandas/core/reshape/tile.py [hunk 12]
```diff
@@ -181,14 +189,45 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
...
+        # GH 14627
+        # NOTE: when the minimum numpy requirement is
+        # increased to 1.15, the try-except statement
+        # can be removed.
+        try:
+            bins = np.histogram_bin_edges(x, bins)
+        except AttributeError:
+            cnt, bins = np.histogram(x, bins)
+
+        mn, mx = bins[0], bins[-1]
```
Comment (TomAugspurger) [inline pandas/core/reshape/tile.py]: Is this equivalent to doing `pd.cut(np.histogram_bin_edges(array, bins))`? Why do we do the additional processing / adjustment starting here?

File: pandas/tests/reshape/test_tile.py [hunk 13]
```diff
@@ -37,6 +36,33 @@ def test_bins(self):
...
+                                              closed="right")
+        expected = Categorical(intervals, ordered=True)
+        tm.assert_index_equal(result.categories,
+                              expected.categories)
+        
+        
+        # Test that a `bin` string not present in `np.histogram_bin_edges`
+        # throws a ValueError.
+        with pytest.raises(ValueError,
+            match="'*' is not a valid estimator for `bins`",
```
Comment (TomAugspurger) [inline pandas/tests/reshape/test_tile.py]: This test will potentially fail if NumPy adjusts their error message. Given that it's not really testing anything in pandas, I think we're OK removing the `match`. It's probably fine to ensure that a ValueError is raised.

**Review by TomAugspurger — COMMENTED** [review]
I think there are some merge conflicts that need to be addressed.

**Comment by bla1089** [comment]
I'm not sure what you're referring to regarding the processing. Are you
asking why the try-except statement? If so, it's because numpy didn't
refactor `histogram_bin_edges` from `histogram` until until v1.15.

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by bla1089** [comment]
Pushing now. I hadn't b/c I was waiting for @TomAugspurger to reply*.*

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by jreback** [comment]
@bla1089 you need to merge master

File: doc/source/whatsnew/v0.24.0.rst [hunk 14]
```diff
@@ -233,6 +233,7 @@ Other Enhancements
 - :class:`DatetimeIndex` gained :attr:`DatetimeIndex.timetz` attribute. Returns local time with timezone information. (:issue:`21358`)
 - :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support an ``ambiguous`` argument for handling datetimes that are rounded to ambiguous times (:issue:`18946`)
 - :meth:`round`, :meth:`ceil`, and meth:`floor` for :class:`DatetimeIndex` and :class:`Timestamp` now support a ``nonexistent`` argument for handling datetimes that are rounded to nonexistent times. See :ref:`timeseries.timezone_nonexistent` (:issue:`22647`)
+- :func:`~cut` `bins` kwarg now accepts a string, which is dispatched to `numpy.histogram_bin_edges`. (:issue:`14627`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.0.rst]: Move to 0.25 at this point

**Review by WillAyd — CHANGES_REQUESTED** [review]
Can you merge master and address conflicts?

File: doc/source/whatsnew/v0.24.0.rst (hunk 14 above)
Comment (blalterman) [reply doc/source/whatsnew/v0.24.0.rst]: I'm unsure how to get my fork to update here. I recently changed my GH id and I'm unsure if that's getting in the way. Any ideas?

**Comment by jreback** [comment]
merge master

**Comment by blalterman** [comment]
That's what I did, which is why I'm confused.

pandas-bla1089 balterma$ git merge master

Already up to date.

pandas-bla1089 balterma$ git push

Everything up-to-date

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by jreback** [comment]
you need to merge upstream
``git merge upstream/master``

**Comment by blalterman** [comment]
pandas-bla1089 balterma$ git merge upstream/master

Already up to date.

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by gfyoung** [comment]
@blalterman : Two questions:

* When you run `git remote -v`, what does `upstream` point to?
* Did you fetch from `upstream` first?

**Comment by blalterman** [comment]
Here's the former.

pandas-bla1089$ git remote -v

origin https://github.com/blalterman/pandas.git (fetch)

origin https://github.com/blalterman/pandas.git (push)

upstream https://github.com/pandas-dev/pandas.git (fetch)

upstream https://github.com/pandas-dev/pandas.git (push)

*fetch upstream* pulled down a bunch of changes. Now what?

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by TomAugspurger** [comment]
Now do a `git merge upstream/master` (and you'll need to fixup the conflicts.

**Comment by blalterman** [comment]
ok...

---------------------
B. L. Alterman
Candidate, Applied Physics
Solar and Heliospheric Research Group
Climate and Space Sciences and Engineering
University of Michigan
balterma@umich.edu

**Comment by WillAyd** [comment]
@blalterman are you able to update here? If you changed your GH credentials might be easiest to push a new PR

**Comment by WillAyd** [comment]
Closing as stale

**Pull request closed without merging**